pocs/utils/rs232.py

```diff
@@ -189,13 +202,16 @@ def get_reading(self):
 
     def clear_buffer(self):
         """ Clear Response Buffer """
-        count = 0
-        while self.ser.inWaiting() > 0:
-            count += 1
-            self.ser.read(1)
-
+        # Not worrying here about bytes arriving between reading in_waiting
+        # and calling reset_input_buffer().
+        count = self.ser.in_waiting
+        self.ser.reset_input_buffer()
```

Comment:
Should this be `reset_output_buffer`? Or perhaps both.

Reply:
The docstring for clear_buffer says "Clear Response Buffer", which I interpret to mean the bytes coming in from the device (i.e. the response to the commands sent by POCS).

There is only a single call to clear_buffer, and it is commented out. Shall I simply remove the clear_buffer function?

Reply:
Actually it is one of the things I was going to test whether we need incorporate more, for input and output. Seems to me that the output from the arduino is lagging behind real-time, in some cases by 10+ seconds. 

In my head the serial line would just output constantly such that whenever you did a read you were getting a current reading but I guess that is not the case (hence, serial, I suppose). Experimenting it seems to work fine to clear the output buffer and then take a reading in order to get as close to "real-time" as possible.

This seems to be the case on input too, where a specified command won't take effect for a number of seconds. By first clearing the input buffer and then sending the command it seems to always be immediate. 

Clearly needs to be explored more and I'm not expert at it so let me know if I need to change how read or if I am misunderstanding something.

Reply:
My experience is that true serial lines need to be read promptly because the internal UART (hardware) buffers are really short. We're dealing with USB to serial here, so those short UART buffers shouldn't be an issue. I don't know what options there are with Linux to get the OS to buffer potentially large amounts of input that the client hasn't read.

I think the thing to do is what I'd suggested in chat at some point: run a separate thread for each serial device "owner" (e.g. the PEAS code for reading from an arduino), so that each can read as fast as possible. For the arduinos, there isn't much need to coordinate between reading and writing, so there could be a thread for reading, and the main thread for the process can do the writing, given that writing is rare. For the mount and dome, we essentially have a synchronous protocol, where we mostly send a command and read a response.